framework/framework.go

```diff
@@ -30,11 +30,66 @@ import (
 )
 
 const (
-	functionStatusHeader = "X-Google-Status"
-	crashStatus          = "crash"
-	errorStatus          = "error"
+	// LegacyCloudEventContextKey exposes the key used to store the legacy cloud event context in the context.Context object.
+	LegacyCloudEventContextKey = legacyCloudEventContextKeyType("LegacyCloudEventContext")
+	functionStatusHeader       = "X-Google-Status"
+	crashStatus                = "crash"
+	errorStatus                = "error"
 )
 
+type legacyCloudEventContextKeyType string
+
+type legacyCloudEvent struct {
+	Context legacyCloudEventContext `json:"context"`
+	Data    interface{}             `json:"data"`
+}
+
+type legacyCloudEventContext struct {
+	EventID   string                  `json:"eventId"`
+	Timestamp string                  `json:"timestamp"`
+	EventType string                  `json:"eventType"`
+	Resource  *map[string]interface{} `json:"resource"`
+}
+
+func (e *legacyCloudEvent) UnmarshalJSON(body []byte) error {
+	event := struct {
+		Context legacyCloudEventContext `json:"context"`
+		Data    interface{}             `json:"data"`
+	}{}
+	err := json.Unmarshal(body, &event)
+	if err != nil {
+		return err
+	}
+
+	// Data is required to be at the top level.
+	if event.Data == nil {
+		err = fmt.Errorf("Missing `data` field in legacy cloud event")
+	} else {
+		e.Data = event.Data
+	}
+
+	// Context can either be a top level `context` field, or the relevant fields themselves at the top level.
+	emptyCtx := legacyCloudEventContext{}
```

Comment:
Might want to avoid using `ctx`, since it sounds like it's a `context.Context`. Unfortunate naming collision.